### PR TITLE
feat(semver): `greaterThanRange()` and `lessThanRange()`

### PR DESCRIPTION
**Pull request by kt3k**
towards #4365

This PR adds `greaterThanRange` and `lessThanRange` APIs without using currently broken `rangeMax` and `rangeMin`.

The test cases are borrowed from `node-semver` (https://github.com/npm/node-semver/tree/main/test/fixtures). They are attributed in the headers of test files

### CODE REVIEW

File: semver/less_than_range_test.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    ["1.0.0 - 2.0.0", "2.0.0"],
+    ["^3.0.0", "4.0.0"],
+    ["^1.0.0 || ~2.0.1", "2.0.0"],
+    ["^0.1.0 || ~3.0.1 || 5.0.0", "3.2.0"],
+    ["^0.1.0 || ~3.0.1 || 5.0.0", "5.0.0-0"],
+    ["^0.1.0 || ~3.0.1 || >4 <=5.0.0", "3.5.0"],
+    ["^1.0.0-alpha", "1.0.0-beta"],
+    ["~1.0.0-alpha", "1.0.0-beta"],
+    ["=0.1.0", "1.0.0"],
+    // Note: Mayve worth considering to add support for includePrerelease option
```
Comment (iuioiua) [inline semver/less_than_range_test.ts:159]:
```suggestion
    // Note: maybe worth considering adding support for `includePrerelease` option
```

File: semver/less_than_range.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+
+import type { Comparator, Range, SemVer } from "./types.ts";
+import { testComparatorSet } from "./_test_comparator_set.ts";
+import { isWildcardComparator } from "./_shared.ts";
+import { compare } from "./compare.ts";
+
+/** Check if the semver is less than the range. */
+export function lessThanRange(semver: SemVer, range: Range): boolean {
```
Comment (iuioiua) [inline semver/less_than_range.ts:9]: Optional for this PR, but full documentation would be nice.

File: semver/less_than_range.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    // If the semver is greater than any of the comparator set, then it's not less than the range.
+    if (
+      comparatorSet.some((comparator) =>
+        greaterThanComparator(semver, comparator)
+      )
+    ) {
+      return false;
+    }
+  }
+  return true;
```
Comment (iuioiua) [inline semver/less_than_range.ts:27]:
```suggestion
  return !range.some((comparatorSet) =>
    comparatorSet.some(isWildcardComparator) ||
    // If the semver satisfies the comparator set, then it's not less than the range.
    testComparatorSet(semver, comparatorSet) ||
    // If the semver is greater than any of the comparator set, then it's not less than the range.
    comparatorSet.some((comparator) =>
      greaterThanComparator(semver, comparator)
    )
  );
```

File: semver/less_than_range.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    }
+    case "<": {
+      return cmp >= 0;
+    }
+    case ">=": {
+      return false;
+    }
+    case "<=": {
+      return cmp > 0;
+    }
```
Comment (iuioiua) [inline semver/less_than_range.ts:54]:
```suggestion
    case "=":
    case undefined:
      return cmp > 0;
    case "!=":
    case ">":
    case ">=":
      return false;
    case "<":
    case "<=":
      return cmp >= 0;
```

File: semver/less_than_range.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        greaterThanComparator(semver, comparator)
+      )
+    ) {
+      return false;
+    }
+  }
+  return true;
+}
+
+function greaterThanComparator(
```
Comment (iuioiua) [inline semver/less_than_range.ts:30]: Nit: This is more of a personal preference, but I like it when functions are defined before they're used. In other words, I'd prefer this function be moved before `lessThanRange()`.

File: semver/greater_than_range.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+      return false;
+    }
+    // If the semver is less than any of the comparator set, then it's not greater than the range.
+    if (
+      comparatorSet.some((comparator) => lessThanComparator(semver, comparator))
+    ) {
+      return false;
+    }
+  }
+  return true;
```
Comment (iuioiua) [inline semver/greater_than_range.ts:25]:
```suggestion
    comparatorSet.some(isWildcardComparator) ||
    // If the semver satisfies the comparator set, then it's not greater than the range.
    testComparatorSet(semver, comparatorSet) ||
    // If the semver is less than any of the comparator set, then it's not greater than the range.
    comparatorSet.some((comparator) => lessThanComparator(semver, comparator))
  );
```

File: semver/greater_than_range.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    }
+    case "<": {
+      return false;
+    }
+    case ">=": {
+      return cmp < 0;
+    }
+    case "<=": {
+      return false;
+    }
```
Comment (iuioiua) [inline semver/greater_than_range.ts:49]:
```suggestion
    case "=":
    case undefined:
      return cmp < 0;
    case "!=":
      return false;
    case ">":
      return cmp <= 0;
    case "<":
      return false;
    case ">=":
      return cmp < 0;
    case "<=":
      return false;
    default:
      return false;
```

**Review by iuioiua — CHANGES_REQUESTED** [review]
Some of my suggestions are optional.

File: semver/less_than_range.ts (hunk 3 above)
Comment (kt3k) [reply semver/less_than_range.ts]: I extracted `lessThanComparatorSet` to make this part more readable

File: semver/greater_than_range.ts (hunk 6 above)
Comment (kt3k) [reply semver/greater_than_range.ts]: I extracted `greaterThanComparatorSet` instead

**Review by iuioiua — APPROVED** [review]

**Pull request merged**